flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -209,6 +211,86 @@ class BoxValue : public AbstractBox, public AbstractArrayBox {
+/// and type are not constant in the entity lifetime (e.g Allocatables and
+/// Pointers).
+class BoxAddressValue : public AbstractBox {
+public:
+  /// Create BoxAddressValue given the address \p addr of the box and the non
+  /// deferred length parameters \p lenParameters. The non deferred length
+  /// parameters must always be provided, even if they are constant and already
+  /// reflected in the address type.
+  BoxAddressValue(mlir::Value addr, llvm::ArrayRef<mlir::Value> lenParameters)
+      : AbstractBox(addr), lenParams{lenParameters.begin(),
+                                     lenParameters.end()} {
+    // Currently only accepts fir.(ref/ptr/heap)<fir.box<type>> mlir::Value for
+    // the address. This may change if we accept
+    // fir.(ref/ptr/heap)<fir.heap<type>> for scalar without length parameters.
+    assert(verify() &&
+           "BoxAddressValue requires mem ref to fir.box<fir.[heap|ptr]<type>>");
+  }
+
+  /// Get the fir.box<type> part of the address type.
+  fir::BoxType getBoxTy() const {
+    auto type = getAddr().getType();
+    return fir::dyn_cast_ptrEleTy(type).cast<fir::BoxType>();
+  }
+  /// Return the part of the address type after memory and box types. That is
+  /// the element type, maybe wrapped in a fir.array type.
+  mlir::Type getBaseTy() const {
+    return fir::dyn_cast_ptrEleTy(getBoxTy().getEleTy());
+  }
+  /// Get the scalar type related to the described entity
+  mlir::Type getEleTy() const {
+    auto type = getBaseTy();
+    if (auto seqTy = type.dyn_cast<fir::SequenceType>())
+      return seqTy.getEleTy();
+    return type;
+  }
+
+  /// Is this a Fortran pointer ?
+  bool isPointer() const {
+    return getBoxTy().getEleTy().isa<fir::PointerType>();
+  }
+  /// Is this an allocatable ?
+  bool isAllocatable() const {
+    return getBoxTy().getEleTy().isa<fir::HeapType>();
+  }
+  /// Is the entity an array or an assumed rank ?
+  bool hasRank() const { return getBaseTy().isa<fir::SequenceType>(); }
+  /// Is this an assumed rank ?
+  bool hasAssumedRank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    return seqTy && seqTy.hasUnknownShape();
+  }
+  /// Returns the rank of the entity. Beware that zero will be returned for
+  /// both scalars and assumed rank.
+  unsigned rank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    if (seqTy)
+      return seqTy.getDimension();
+    return 0;
+  }
+  /// Is this a character entity ?
+  bool isCharacter() const { return getEleTy().isa<fir::CharacterType>(); };
+  /// Is this a derived type entity ?
+  bool isDerived() const { return getEleTy().isa<fir::RecordType>(); };
+  /// Does this entity has any non deferred length parameters ?
+  bool hasNonDeferredLenParams() const { return !lenParams.empty(); }
+  /// Return the non deferred length parameters.
+  llvm::ArrayRef<mlir::Value> nonDeferredLenParams() { return lenParams; }
```

Comment:
"non deferred length parameters" exactly matches Fortran usage (e.g in 9.7.1.2 point 9), i.e it is a length type parameters that is not deferred (`:`) in the allocatable declaration. That means is either assumed or it is a specification expressions (dynamic or compile time constant). In any case, it means the length is constant in the scope being lowered and cannot be changed by any allocate statement. If non deferred, length parameters can actually be omitted from the allocate statement. That's why we need to keep track of them.

```
! deferred
character(:), allocatable :: c1
! non deferred: specification expression (compile time constant)
character(10), allocatable :: c2
! non deferred: specification expression (dynamic)
character(n), allocatable :: c3
! non deferred: assumed length (dynamic), only possible if c4 is a dummy
character(*), allocatable :: c4

! Case 1: OK
allocate(c2, c3, c4)

! Case 2: Not OK, the length that was deferred is not provided. Caught by f18 semantics checks 
allocate(c1)

! Case 3: OK
allocate(character(l):: c1)

! Case 4: OK, as long as l does not conflict with the non deferred length (can only be checked at runtime)
allocate(character(l):: c2, c3, c4)
```

To lower Case 1, we know the length that was specified in the declaration. It is too late to lower the length from the symbol at this point because if it is a dynamic expression using symbols, the value may have change (e.g. `n` may have changed for c3 between the specification part and the allocate statement ). So we need to remember the non deferred length, at least when this is a dynamic specification expression.

In Case 4, we have the opportunity to insert a runtime check that l matches the non deferred length parameters.


This will also be useful for length parameters inquires:
```
!  length inquiry on unallocated is OK for non deferred length parameters (9.4.5 point 2)
print *, c2%LEN, c%LEN,  c4%LEN
! only OK if c1 is allocated. Runtime check opportunity.
print *, c1%LEN 
```
Here again, having the list of non deferred length parameters should help.

I did not dealt properly yet with the assumed case (I am not adding from the input descriptor the length to this list).

Note that bounds are necessarily deferred for allocatables, so all that only really matters for character and derived type with length parameters.